File: src/code/go/03_stack/01_new_stack/linked_list_stack.go
package new_stack

type node struct {
	val  any
	next *node
}

var _ Stack = (*LinkedListStack)(nil)

type LinkedListStack struct {
	top *node
}

func NewLinkedStack() *LinkedListStack {
	return &LinkedListStack{}
}

// Push 入栈
func (l *LinkedListStack) Push(val any) {
	l.top = &node{val: val, next: l.top}
}

// Pop 出栈
func (l *LinkedListStack) Pop() any {
	if l.IsEmpty() {
		return nil
	}

	node := l.top
	l.top = l.top.next
	return node.val
}

// IsEmpty 是否为空
func (l *LinkedListStack) IsEmpty() bool {
	return l.top == nil
}

// Top 置顶元素
func (l *LinkedListStack) Top() any {
	if l.IsEmpty() {
		return nil
	}
	return l.top.val
}

// Flush 清空
func (l *LinkedListStack) Flush() {
	l.top = nil
}

File: src/code/php/03_stack/01_new_stack/linked_list_stack.php
<?php

require_once "./stack_interface.php";

class LinkedListNode
{
    public function __construct(
        public mixed $val = 0,
        public LinkedListNode|NULL $next = null
    ) {
    }
}

class LinkedListStack implements StackInterface
{
    /** LinkedNode 栈顶结点 */
    public LinkedListNode|NULL $top = null;

    /**
     * 添加元素
     *
     * @param mixed $val
     * @return void
     */
    public function push(mixed $val)
    {
        $node = new LinkedListNode($val, $this->top);
        $this->top = $node;
    }

    /**
     * 弹出元素
     *
     * @return mixed
     */
    public function pop(): mixed
    {
        if ($this->isEmpty()) {
            return null;
        }

        $node = $this->top;
        $this->top = $this->top->next;
        return $node->val;
    }

    /**
     * 判断是否为空
     *
     * @return boolean
     */
    public function isEmpty(): bool
    {
        return $this->top == null;
    }

    /**
     * 清空栈
     *
     * @return void
     */
    public function flush()
    {
        $this->top = null;
    }

    /**
     * 返回栈顶元素
     *
     * @return mixed
     */
    public function top(): mixed
    {
        if ($this->isEmpty()) {
            return null;
        }

        return $this->top->val;
    }
}

$stack = new LinkedListStack();
$stack->push(1);
$stack->push("a");
print($stack->top());
print($stack->pop());
print($stack->pop());
